### ISSUE TITLE
Scrolling with two fingers scrolls twice as fast

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#11884]
When scrolling a Flutter list with two fingers the list scrolls twice as fast.

In native Android or iOS apps, the list continues to scroll in regular speed even with two fingers.

Note: This is NOT specific to accessibility scrolling with two fingers. It's visible even when accessibility is disabled. However, it's also a problem for accessibility scrolling, where using two fingers are commonly used for scrolling.

**Comment by eseidelGoogle** [comment]
While this is awesome, I suspect this will block AX work and we should remove it. :)

**Comment by eseidelGoogle** [comment]
I'm not able to repro this anymore, presumably this got fixed?

**Comment by tevjef** [comment]
The scrolling speed increases with the number of fingers on screen.

**Comment by wanjm** [comment]
I still come across with such problem with 1.9.1 +hotfix4 & 5;
any progress on this?

**Comment by VladyslavBondarenko** [comment]
Still persists (master v1.16.3-pre.56)

**Comment by dkwingsmt** [comment]
Part of: https://github.com/flutter/flutter/issues/38926

**Comment by VladyslavBondarenko** [comment]
Persists with current master `1.21.0-6.0.pre.38`

<details>
 <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 1.21.0-6.0.pre.38, on Mac OS X 10.15.6 19G73, locale en-GB)
    • Flutter version 1.21.0-6.0.pre.38 at /Users/nevercode/dev/flutter
    • Framework revision 8e0eee9008 (5 hours ago), 2020-07-26 23:38:01 -0700
    • Engine revision 626244a72c
    • Dart version 2.9.0 (build 2.9.0-21.0.dev a3815b6590)
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/nevercode/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.47.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.2

[✓] Connected device (5 available)
    • Redmi Note 7 (mobile)       • 6345c469                                 • android-arm64  • Android 9 (API 28)
    • Nevercode’s iPhone (mobile) • b668e524315069f3db3661ac11ff1f66afafebdb • ios            • iOS 13.6
    • macOS (desktop)             • macos                                    • darwin-x64     • Mac OS X 10.15.6 19G73
    • Web Server (web)            • web-server                               • web-javascript • Flutter Tools
    • Chrome (web)                • chrome                                   • web-javascript • Google Chrome 84.0.4147.89

• No issues found!
```

</details>

**Comment by fzyzcjy** [comment]
I find a workaround for it! :D 

Please look at this [question](https://stackoverflow.com/questions/64220989/listview-only-allow-double-finger-scrolling-and-disable-single-finger-one-flu/64222689#64222689) (answered by myself after digging into flutter source code).

Note: Do not worry about that question's title. If you set `onlyScrollableWhenMultiFinger: true` inside my sample shown there, you will solve this question, without disabling single finger scrolling.

**Comment by adschi** [comment]
Any solution for the whole app to disable that feature?

**Comment by RandalSchwartz** [comment]
Can we leave something like this in there though?  It's a great shibboleth to detect Flutter apps in the wild.

**Comment by UnicornsOnLSD** [comment]
There are plenty of ways to tell whether or not an app uses Flutter, I'm sure we can fix this one 🙃

**Comment by RandalSchwartz** [comment]
> There are plenty of ways to tell whether or not an app uses Flutter, I'm sure we can fix this one 🙃

Like?

**Comment by jmshrv** [comment]
Off the top of my head:

* On iOS, the scrollbar stops when you try to flick it. In native iOS apps, the scrollbar keeps momentum after you let go
* The inkwell effect in Material buttons fills slowly. In native Android apps, it fills within ~0.25 seconds
* On iOS, holding down the back button doesn't show the list of previous pages (I haven't tested this with CupertinoApp so apologies if it works there)
* When scrolling a ScrollView that isn't big enough to actually scroll, the overflow effect will show in Android

**Comment by alxlion** [comment]
> > There are plenty of ways to tell whether or not an app uses Flutter, I'm sure we can fix this one 🙃
> 
> Like?

Like [FlutterHunt](https://flutterhunt.com) I think

**Comment by alxlion** [comment]
I'm trying to solve this issue, not easy but I found this: 
[`flutter/lib/src/gestures/monodrag.dart:274`](https://github.com/flutter/flutter/blob/5a279edc239f4efd91223ddc609c2b5129325feb/packages/flutter/lib/src/gestures/monodrag.dart#L267)

```
@override
  bool isPointerAllowed(PointerEvent event) {
    if (_initialButtons == null) {
      if (onDown == null && onStart == null && onUpdate == null && onEnd == null && onCancel == null) {
        return false;
      }
    } else {
      // There can be multiple drags simultaneously. Their effects are combined.
      if (event.buttons != _initialButtons) {
        return false;
      }
    }
    return super.isPointerAllowed(event as PointerDownEvent);
  }
```

> There can be multiple drags simultaneously. Their effects are combined.

It seems to be a normal behavior, but why ?

**Comment by gnprice** [comment]
> > There can be multiple drags simultaneously. Their effects are combined.
> 
> It seems to be a normal behavior, but why ?

Interesting find!

I wouldn't necessarily read that comment as meaning that the behavior is "normal", though, or even that anyone at any point actually intended it. The comment could instead be something added by someone who discovered that was the behavior, and wanted to write it down to help explain the working of the new code they were adding.

In fact, looking in the Git history (`git log -p -S 'effects are combined' lib/src/gestures/monodrag.dart`), the comment was first added in #30339 in 2019. That's nearly two years after this issue was filed, so the behavior was already there.

**Comment by gnprice** [comment]
I took a quick look at how other platforms seem to handle this situation. In particular, when two fingers are dragging on something scrollable it's a bit ambiguous how the scrollable should respond, because the fingers may not be moving precisely in tandem.

The procedure was that I opened up the system settings app in iOS and Android respectively, scrolled up and down the list using two fingers, and experimented with different ways of moving the two fingers.
* Android: Whichever finger was the last to start touching the screen, that finger controls the motion. The earlier finger can stay still, or move around wildly, and it makes no difference; the scrollable follows the last finger.

  OTOH if finger A starts dragging, then finger B touches down and takes over, then finger B lifts, the result is that A once again controls.

* iOS: If the fingers are moving in the same direction, then whichever one is moving fastest controls the motion.

  If the fingers are moving in opposite directions… then the scrollable doesn't move very much, but occasionally jerks around in small motions, feeling a bit glitchy. I can't quite tell what the exact behavior is.

  It's possible that the rule is intended to be that if they move in opposite directions, then the scrollable doesn't move at all. The trouble with that rule is that its behavior is highly discontinuous — if finger A is moving fast and finger B moves slightly in the same direction, that has a very different result from if finger B moves slightly in the opposite direction, or doesn't move — and the jerkiness could be a symptom of that, given the fact that my finger doesn't move with perfect smoothness and the touch screen may be imperfect too.

The Android rule is unsatisfyingly asymmetric. But OTOH I feel like I've regularly felt as an iOS user that scrolling gets a little glitchy when another finger touches the screen, and as an Android user I haven't felt that way. Given that plus the clearly glitchy-feeling behavior iOS has when pressing things to an edge case like I did just now, the ideal thing might be to implement the Android latest-finger-wins behavior and use that on all platforms. Or at least to start with that and see if there's any desire for other behaviors once that's done.

**Comment by darshankawar** [comment]
Verified this issue using latest versions and observed the fast scrolling using two fingers as below:

https://user-images.githubusercontent.com/67046386/224007908-11b4f4bf-fcb8-4bed-8543-d36c8a7626d7.MP4

Code sample used: https://github.com/flutter/flutter/issues/11884#issuecomment-605050113

<details>
<summary> stable, master flutter doctor -v </summary>

```
[!] Flutter (Channel stable, 3.7.5, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 3.7.5 on channel stable at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision c07f788888 (7 days ago), 2023-02-22 17:52:33 -0600
    • Engine revision 0f359063c4
    • Dart version 2.19.2
    • DevTools version 2.20.1
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

[!] Flutter (Channel master, 3.9.0-1.0.pre.118, on macOS 12.2.1 21D62
    darwin-x64, locale en-GB)
    • Flutter version 3.9.0-1.0.pre.118 on channel master at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 036592e0cf (38 minutes ago), 2023-03-08 23:15:14 -0500
    • Engine revision 7e1b3a98d1
    • Dart version 3.0.0 (build 3.0.0-312.0.dev)
    • DevTools version 2.22.2
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/dhs/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for
      more details.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13C100
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 65.1.4
    • Dart plugin version 213.7228

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.29.0

[✓] Connected device (3 available)
    • Darshan's iphone (mobile) • 21150b119064aecc249dfcfe05e259197461ce23 • ios
      • iOS 15.3.1 19D52
    • macOS (desktop)           • macos                                    •
      darwin-x64     • macOS 12.2.1 21D62 darwin-x64
    • Chrome (web)              • chrome                                   •
      web-javascript • Google Chrome 109.0.5414.119

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
      
[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.



```

</details>

**Comment by hitiksaini** [comment]
This should be left unsolved(only if this doesn't break things elsewhere).
Normally, I use it as a quick way to check if an app is using flutter 😄

**Comment by JanFeni** [comment]
@hitiksaini  This does break the user experience in a common use case though.

**How?**
 If you have 2 or more fingers scrolling in opposite directions, they sort of cancel each other out but not quite, ending in some sort of janky movement going either direction.

**The issue**
The issue is that, and this is rather apparent on low to low-medium-range phones I have tested with rounded edges or borderless screens, the palm of the hand is mistaken by the device as a single or many fingers, therefore during one-hand use of a device, the palm of the hand touching the border of the screen interfere with the actual scroll of the user, sometimes not by much (slowing down the actual scroll "randomly"), sometimes quite badly. Even when it is minimal, it is noticeable enough to break the experience. 

Medium and higher-end phones do not seem to have this issue with hand palms however this still should be fixed so as to not alienate an entire market.

**Comment by hitiksaini** [comment]
> @hitiksaini This does break the user experience in a common use case though.
> 
> **How?** If you have 2 or more fingers scrolling in opposite directions, they sort of cancel each other out but not quite, ending in some sort of janky movement going either direction.
> 
> **The issue** The issue is that, and this is rather apparent on low to low-medium-range phones I have tested with rounded edges or borderless screens, the palm of the hand is mistaken by the device as a single or many fingers, therefore during one-hand use of a device, the palm of the hand touching the border of the screen interfere with the actual scroll of the user, sometimes not by much (slowing down the actual scroll "randomly"), sometimes quite badly. Even when it is minimal, it is noticeable enough to break the experience.
> 
> Medium and higher-end phones do not seem to have this issue with hand palms however this still should be fixed so as to not alienate an entire market.

@JanFeni I understand your point of view. I haven't got my hands on with low-end phones. Thanks for this information.

**Comment by milindgoel15** [comment]
It could become an optional feature tho. Like making a Boolean at the top level say MaterialApp (in case of android) as `enableTwoFingerFastScroll` or something and keeping it off by default.

**Comment by definev** [comment]
This issues is dramatically affecting to many app that have Tiktok-like vertical video, if user scrolling using 2 finger it will unconditionally move faster so very bad UX

**Comment by xu-baolin** [comment]
Hahahaha, so interesting

**Comment by 2p31-1** [comment]
Still exists on my SingleChildScrollView

**Comment by JacobSyndeo** [comment]
Following up on @gnprice's comment regarding scroll behavior, Apple platforms average the motion of all fingers, rather than summing them.

Try scrolling with two fingers, then leaving one in place while continuing to move the second. You'll notice the motion is now halved. Now pause the one that you kept moving and moved the first one in the other direction. You'll see it moves back at half speed as well. Now move both fingers towards each other, and then back apart again; you shouldn't see much scroll motion at all.

It essentially creates a virtual scroll "touch" at the average/midpoint between the two fingers. And I believe that this implementation that would be a great solution here as well, at least when targeting Apple platforms.

If Android's behavior differs, listening only to the most recent touch, then I believe Flutter should do the same when targeting Android.

There's definitely something to be said for having consistent behavior across platforms (and this is one of Flutter's strengths), but scrolling feels like such a fundamental part of the UX that I believe it should be tailored to each platform's stock behavior by default.

But in the spirit of full customization, (while this runs counter to good UX IMO), if certain developers really want to set a specific behavior across their app regardless of platform, we can expose this as an enum option.

Something like this:

```Dart
enum MultitouchScrollBehavior { auto, average, first, last, cumulative }
```

Where `auto` uses `average` for Apple and `last` for Android. We could add `first` for new behavior, but I'm not certain of the value. Finally, `cumulative` (maybe called `sum` instead?) would be the current existing behavior.

Whether or not we want to leave existing projects on `cumulative` or default them to `auto` moving forward is another discussion, IMO.

**Comment by JacobSyndeo** [comment]
Wow, would you look at that… I feel like I'm in the twilight zone or something, because I could've sworn that's how it worked.

In any case, I stand corrected; thanks for pointing that out, @goderbauer.

(Unfortunate, too… I was _**just now**_ opening a PR that included a behavior fix for both iOS _and_ Android, with the expected behaviors. I'll have to refactor it!)

**Comment by squirelboy360** [comment]
This should be taken off immediately. We are in the era where flutter apps are almost equivalent to native in terms of performance, we chose to learn flutter not for it to dictate the respective OS how it is supposed to scroll but rather vice versa. Doing this and telling developers it is a trail left to show users our app is a flutter signature app is in my opinion dumb. Like why. There should be a property to disable this like the way we could disable material 3.

**Comment by jmshrv** [comment]
Take a look at the PR - `MultiTouchDragBehavior.sum` is available if you wanted the old scrolling for some reason. [The Flutter docs has an interesting page on how it adapts to the platform](https://docs.flutter.dev/platform-integration/platform-adaptations) if you're interested.

**Comment by squirelboy360** [comment]
Really🤓, i did not know cause my apps after flutter upgrade are still scrolling in two or three times as fast depending on the number of fingers on screen.

**Comment by esentis** [comment]
@squirelboy360 It's not on stable yet, it will be released with 3.18.0 version.

**Comment by gnprice** [comment]
For the sake of making the outcome visible explicitly on this thread: the current solution is that the default behavior on all platforms is that the latest finger controls the scrolling. That's the same as the Android native behavior, as described at https://github.com/flutter/flutter/issues/11884#issuecomment-1461343091.

See https://github.com/flutter/flutter/pull/136708 which made the fix, and the docs at:
https://main-api.flutter.dev/flutter/gestures/DragGestureRecognizer/multitouchDragStrategy.html
https://main-api.flutter.dev/flutter/gestures/MultitouchDragStrategy.html

(I figured this was worth clarifying here because the latest discussion above was toward a different option, where we'd also implement the iOS native behavior and make that the default on iOS. That may still happen in the future if there's a desire for it, but it'd be its own issue.)

Very happy this is fixed! Thanks @xu-baolin @JacobSyndeo and everyone else who contributed.
